src/include/ipc/dai.h

```diff
@@ -61,30 +61,42 @@ enum sof_ipc_dai_type {
 	SOF_DAI_INTEL_ALH,		/**< Intel ALH */
 	SOF_DAI_IMX_SAI,                /**< i.MX SAI */
 	SOF_DAI_IMX_ESAI,               /**< i.MX ESAI */
+	SOF_DAI_MULTIDAI = 0xFFFF       /**< config with multiple elements of the same dai type */
 };
 
 /* general purpose DAI configuration */
 struct sof_ipc_dai_config {
 	struct sof_ipc_cmd_hdr hdr;
-	uint32_t type;		/**< DAI type - enum sof_ipc_dai_type */
+
+	/* FIXME endianness */
+	uint16_t type;		/**< DAI type - enum sof_ipc_dai_type */
+	uint16_t subdai_type;	/**< only used in multi-dai cases */
 	uint32_t dai_index;	/**< index of this type dai */
 
 	/* physical protocol and clocking */
 	uint16_t format;	/**< SOF_DAI_FMT_ */
-	uint16_t reserved16;	/**< alignment */
+
+	/* we can either have multiple sub-DAIs or multiple configs, not both */
+	uint16_t num_payloads; 	/**< number of configs or number of sub-DAIs */
 
 	/* reserved for future use */
 	uint32_t reserved[8];
 
 	/* HW specific data */
 	union {
-		struct sof_ipc_dai_ssp_params ssp;
-		struct sof_ipc_dai_dmic_params dmic;
-		struct sof_ipc_dai_hda_params hda;
-		struct sof_ipc_dai_alh_params alh;
-		struct sof_ipc_dai_esai_params esai;
-		struct sof_ipc_dai_sai_params sai;
+		struct sof_ipc_dai_ssp_params ssp[0];
+		struct sof_ipc_dai_dmic_params dmic[0];
+		struct sof_ipc_dai_hda_params hda[0];
+		struct sof_ipc_dai_alh_params alh[0];
+		struct sof_ipc_dai_esai_params esai[0];
+		struct sof_ipc_dai_sai_params sai[0];
 	};
+
+	/* multi-dai dai index variable array */
+	uint32_t subdai_index[0];
+
+	/* multi-dai stream maps. The ext_id used need to match subdai_index */
+	struct sof_ipc_stream_map stream_map;
```

Comment:
I honestly don't like having two variable (0-length) arrays and then the stream_map member. What is this exactly supposed to mean?

Reply:
Yeah, all members past the flex arrays will have to be serialized, they won't be able to be accessed directly.

Reply:
yeah, we really have 3 flex arrays here since stream_map contains one too.
Is there a 1:1 mapping between subdai_index size ? If so, we could combine the array in order to simplify.

Reply:
we could add the subdai_index to all existing params definitions. that way in multi-dai configuration you'd know exactly which payload maps to what subdai_index.

But for the stream_map I have no solution. It's already defined as a separate variable array so no matter what the firmware will have to do a mapping between the ext_id and the subdai_index.

Reply:
ack, lets address in IPC2.0

Reply:
@plbossart I could be completely offtrack here. But can we maybe move the sof_ipc_channel_map field in the stream map to the alh_params and get rid of the stream_map field above? It looks like num_ch_map field should be the same as the num_payloads right?

Reply:
> @plbossart I could be completely offtrack here. But can we maybe move the sof_ipc_channel_map field in the stream map to the alh_params and get rid of the stream_map field above? It looks like num_ch_map field should be the same as the num_payloads right?

@ranj063 I think we can't move sof_ipc_channel_map to alh_params since multi-dai is not for ALH only. For example, we could also use multi-dai for SSP though we usually use TDM.  My point is that it is better to keep the capability of supporting all kinds of  dais though we don't need it for now.

Reply:
@bardliao yes I understand that it may not be alh only. In that case we should probably add it to the respoective params whenever needed. 

Alternatively, may I suggest we move the stream_map before the union of flex arrays? That way, at least all the flex arrays will be at the end?

Reply:
the stream map contains a flex array.

Reply:
@plbossart yes, I saw that. Thats why I am thinking that if we move stream map before the union, all the flex arrays will be at the end of the structure.

Reply:
it'd make the 'standard' case more complicated and break backwards compatibility